std/random.d

```diff
@@ -1336,6 +1330,271 @@ version (AnyARC4Random)
     extern(C) private @nogc nothrow
     {
         uint arc4random() @safe;
+        void arc4random_buf(scope void* ptr, size_t len) @system;
+    }
+
+    /+
+    Tries to use a system API to get entropy. Resources are not cached
+    due to the expectation that this function will not be called more
+    than once per program execution.
+
+    Params:
+        ptr = address at which to start writing
+        len = number of bytes to write
+
+    Returns:
+        always `len` (does not fail)
+    +/
+    private ptrdiff_t oneTimeGenRandom()(scope void* ptr, ptrdiff_t len) @system
+    {
+        arc4random_buf(ptr, len);
+        return len;
+    }
```

Comment:
>There is also a separate point of consideration about whether a function to get system entropy should default to `arc4random`. This is essentially a judgement call phobos-side that arc4random is preferable to any other implementation. Is that necessarily true? And if so, why?

The main attractive feature of `arc4random` is its simplicity. It requires no Phobos-side management of global state and has no error conditions. It is hard to imagine a better interface.

>why not define a function which provides entropy generated using the standard entropy-generating mechanism of the host system, whatever that may be?

For the most part we do that.

Reply:
> The main attractive feature of `arc4random` is its simplicity. It requires no Phobos-side management of global state and has no error conditions. It is hard to imagine a better interface.

That's obviously nice to have, but I don't know that I'd necessarily privilege that over other factors, like the quality of the source of entropy.

> For the most part we do that.

I accept that this is the intent, but it's not really what the _code_ does, because too much stuff is hidden under the hood in a way that denies choices to the actual use-cases.

Consider what you have right now:

```D
version (AnyARC4Random)
{
    private ptrdiff_t oneTimeGenRandom()(scope void* ptr, ptrdiff_t len) @system
    {
        // arc4random-based implementation
    }
}
else version (Windows)
{
    private ptrdiff_t oneTimeGenRandom()(scope void* ptr, ptrdiff_t len) @system
    {
        // Windows syscall based implementation
    }
}
else version (Posix)
{
    private ptrdiff_t oneTimeGenRandom()(scope void* ptr, size_t len) @system
    {
        // Posix syscall based implementation, with some scope for Linux specialization
    }
}
else
{
    // implementation that always fails
}
```

Contrast this with:

```D
// you already have this ...
version (AnyARC4Random)
{
    extern (C) @private @nogc nothrow
    {
        uint arc4random() @safe;
        void arc4random_buf(scope void* ptr, size_t len) @system;
    }
}

// ... but you don't have this
version (linux)
{
    int getrandom(void* ptr, size_t len, out bool isImplemented)
    {
        // wrap up all the syscall stuff inside this
    }
}

// ... or this
version (Posix)
{
    int devURandom(void ptr*, size_t len)
    {
        // wrap up the implementation here
    }
}

// ... or this
version (Windows)
{
    int fnCryptGenRandom(void ptr*, size_t len)
    {
        // wrap up implementation here
    }
}

// ... and now, you can do this
int getEntropy(void ptr*, size_t len)
{
    version (AnyARC4Random)
    {
        arc4random_buf(ptr, len);
        return len;
    }
    else version (Windows)
    {
        return fnCryptGenRandom(void ptr*, size_t len);
    }
    else version (Posix)
    {
        version (linux)
        {
            bool getrandomExists;
            result = getrandom(ptr, len, getrandomExists);
            if (getrandomExists) return result;
        }

        return devURandom(ptr, len);
    }
    else
    {
        return -1;
    }
}
```

What's the difference?

With the first option, you have one single "get some entropy from the system" function.  It's a black box.  It may interact in odd ways with other choices made elsewhere in the codebase (as indeed this one does, in the case of `AnyARC4Random` and `unpredictableSeed`).

With the second option, there's still a general-purpose "get system entropy" function, but it's defined as an aggregate of lots of other choices.  It doesn't preclude a user making use of those choices in a different way, depending on their use-case.

For example, with this way of doing things, it would be trivial, if we wanted to, to switch out the `unpredictableSeed` implementation to include:

```D
version (Windows)
{
    UIntType result = void;
    // notice, we're using a _specific_ source of entropy here, not just some
    // abstracted-away black box
    auto bytesOfEntropy = fnCryptGenRandom(&result, UIntType.sizeof);
    enforce(bytesOfEntropy == UIntType.sizeof);
    return result;
}
else version (AnyARC4Random)
{
    // ... existing implementation continues from here ...
}
```

Of course, we might not want to make that specific change.  The key thing is to note how this enables `unpredictableSeed` to make its own choices with its own priorities, that need not necessarily be the same as `getEntropy`.

Reply:
That's a fine way to structure it. That's similar to how `mir-random` divides things.